### PR TITLE
v5.2.3 - Set Virgil-Agent header

### PR DESCRIPTION
**Pull request by xlewkanx**
### **React Native**

Screen Shot 2019-04-25 at 6 53 12 PM
Screen Shot 2019-04-25 at 6 50 44 PM

### **Ionic**

Screen Shot 2019-04-25 at 6 51 49 PM
Screen Shot 2019-04-25 at 6 44 40 PM

### **Usage from other products**

Screen Shot 2019-04-25 at 7 18 19 PM
```
        this.cardManager = new CardManager({
            cardCrypto: this.cardCrypto,
            cardVerifier: this.cardVerifier,
            accessTokenProvider: this.accessTokenProvider,
            retryOnUnauthorized: true,
            connection: new Connection(options.apiUrl, new VirgilAgent('e3kit', packageJson.version)),
        });
```

### CODE REVIEW

File: scripts/rollup/get-rollup-plugins.js [hunk 1]
```diff
@@ -25,15 +29,14 @@ function getRollupPlugins (bundleType) {
 			typescript: require('typescript')
 		}),
 
-		replace({ 'process.browser': JSON.stringify(isBrowser) }),
+		replace({ 'process.env.browser': JSON.stringify(isBrowser) }),
```
Comment (vadimavdeev) [inline scripts/rollup/get-rollup-plugins.js]: Why did you change this line?

File: src/Client/VirgilAgent.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import { OS_LIST } from '../Lib/platform/osList';
+import { BROWSER_LIST } from '../Lib/platform/browserList';
+
+/**
+ * Class responsible for tracking basic information which Virgil Product is using, his version,
```
Comment (vadimavdeev) [inline src/Client/VirgilAgent.ts]:
```suggestion
 * Class responsible for tracking which Virgil SDK is being used to make requests, its version,
```

File: src/Client/VirgilAgent.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	private getUserAgent(): string {
+		if (typeof navigator !== "undefined" && typeof navigator.userAgent === "string") {
+			return navigator.userAgent;
+		} else {
+			return "";
+		}
+	}
+
+	/**
+	 * Initializes a new instance of `Connection`.
```
Comment (vadimavdeev) [inline src/Client/VirgilAgent.ts]:
```suggestion
	 * Initializes a new instance of `VirgilAgent`.
```

File: src/Client/VirgilAgent.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	 * @param {string} product - name of product eg (sdk, brainkey, bpp, keyknox, ratchet, e3kit, purekit)
+	 * argument of request methods.
+	 * @param {string} version - version of the product.
+	 * @param {string} userAgent - string with device user agent. Optional
```
Comment (vadimavdeev) [inline src/Client/VirgilAgent.ts]: Optional parameters names must be wrapped in brackets `[userAgent]`

File: src/Client/VirgilAgent.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+		this.value = `${product};js;${this.getHeaderValue()};${version}`;
+	};
+
+	/**
+	 * Detects device OS
+	 * @returns {string} returns OS if detected or 'other'.
+	 */
+	getOsName() {
+		const os = OS_LIST.find((os) => {
+			if (os.test.some(condition => condition.test(this.userAgent))) return true;
```
Comment (vadimavdeev) [inline src/Client/VirgilAgent.ts]: why not just:
```suggestion
			return os.test.some(condition => condition.test(this.userAgent)))
``` 
`some` already returns boolean

File: src/Client/VirgilAgent.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+		return os ? os.name : 'other';
+	}
+
+	/**
+	 * Detects device browser
+	 * @returns {string} returns browser if detected of 'other'.
+	 */
+	getBrowser() {
+		const browser = BROWSER_LIST.find((browser) => {
+			if (browser.test.some(condition => condition.test(this.userAgent))) return true;
```
Comment (vadimavdeev) [inline src/Client/VirgilAgent.ts]: Same comment as in `getOsName` applies here

File: src/Client/VirgilAgent.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+	 */
+	private isReactNative(): boolean {
+		return typeof navigator === "object" && navigator.product === "ReactNative";
+	}
+
+	/**
+	 * Detect Node.js.
+	 * @return {boolean} true if detects Node.js.
+	 */
+	private isNodeSdk = (): boolean => !process.env.browser && !!process.platform;
```
Comment (vadimavdeev) [inline src/Client/VirgilAgent.ts]: Doesn't this cause the inclusion of `process` polyfill in browsers? If not, then in browsers this must be causing `ReferenceError: process is not defined`. If yes, then we need another way to check for node.js

Comment (vadimavdeev) [inline src/Client/VirgilAgent.ts]: Why the `Sdk` suffix? Other methods don't have it: `isReactNative`, `isIonic`

File: src/index.ts [hunk 8]
```diff
@@ -6,6 +6,8 @@ export * from './Cards/CardManager';
 export * from './Cards/CardVerifier';
 export * from './Cards/RawSignedModel';
 export * from './Cards/ModelSigner';
+export * from './Client/Connection';
+export * from './Client/VirgilAgent';
```
Comment (vadimavdeev) [inline src/index.ts:9]: I'd really like to avoid exporting these classes. They don't add much value to consumers of SDK but complicate the API and should thus be "internal" in my opinion. There must be other ways to use a different header in e3kit (not sure it's even necessary)

File: scripts/rollup/get-rollup-plugins.js (hunk 1 above)
Comment (xlewkanx) [reply scripts/rollup/get-rollup-plugins.js]: There was some typing problems

File: src/Client/VirgilAgent.ts (hunk 7 above)
Comment (xlewkanx) [reply src/Client/VirgilAgent.ts]: this value is replacing during compile time

File: src/index.ts (hunk 8 above)
Comment (xlewkanx) [reply src/index.ts:9]: I think we don't use one of the main feature of OOP - reusability. All classes can be used independently. And I show the example how I use them in e3kit. `Connection` class could be used in keyknox, but not used because it wasn't exported. And we need to detect keyknox and e3kit also.
I agree that VirgilAgent could not be exported, because it can be encapsulated in `Connection` class.
About consumers, we make products based on Virgil SDK, so why not to export the `classes`which we can reuse. And about complexity of SDK it need to be handled by documentation. We provide high level options for consumers and documented them, so there is official way how to use them. And if class is exported it doesn't mean it need to be used by everyone.

**Comment by xlewkanx** [comment]
Changed CardManager initialization to:
```
       this.cardManager = new CardManager({
            cardCrypto: this.cardCrypto,
            cardVerifier: this.cardVerifier,
            accessTokenProvider: this.accessTokenProvider,
            retryOnUnauthorized: true,
            productInfo: {
                product: 'e3kit',
                version: '0.3.7',
            },
        });
```

File: package.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "virgil-sdk",
-  "version": "5.2.2",
+  "version": "5.2.3",
   "description": "Virgil Security Services SDK",
   "contributors": [
```
Comment (vadimavdeev) [inline package.json:5]: Add yourself to contributors 😉

**Review by vadimavdeev — DISMISSED** [review]
👍

**Review by vadimavdeev — APPROVED** [review]
